### PR TITLE
#20877: Update BN with hash function

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/20877

### Problem description
While enabling program cache for the Pytorch project, `ttnn.batch_norm` gives low PCC while testing followed by `ttnn.floor_divide`

### What's changed
Added the missing hash function solved this and also tested the program that caused the issue

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/14663485397) CI passes
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/14663486662) CI passes
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/runs/14663487929) CI passes as in main
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/runs/14663489433)  CI passes as in main
- [x] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/runs/14663491169)  CI passes as in main

### CODE REVIEW

**Review by jmalone-tt — COMMENTED** [review]
Change looks good to me! I'll leave it up to one of the code owners to approve since I haven't looked at many OP hashes yet

**Review by bbradelTT — APPROVED** [review]

**Review by yugaoTT — APPROVED** [review]
lgtm, just a question why is the default hash not working? I thought it should capture everything you passed to program config, and the BN api.

**Review by edwinleeTT — APPROVED** [review]

**Comment by nsorabaTT** [comment]
> lgtm, just a question why is the default hash not working? I thought it should capture everything you passed to program config, and the BN api.

+1 will be good to know why default hash does not work in the PR description.

**Review by nsorabaTT — APPROVED** [review]

**Review by mcw-anasuya — APPROVED** [review]

**Comment by VirdhatchaniKN** [comment]
Hi @yugaoTT , @nsorabaTT 
Are there any disadvantages in using a custom hash function?

**Comment by yugaoTT** [comment]
Using customized hash is fine if you want customized feature, but here doesn't seem like it?
I guess the problem here is when you change tensor layout/dtype it keeps using the previous program cache?
What I see in the program factory is it's not using ProgramWithCallbacks which can update your tensor address.

for example in softmax, it is using this callback that when you hit the cache, it can invoke this function and update your tensor addresses.
```

auto override_runtime_arguments_callback =
        [reader_kernels_id, cb_in0_id, cb_out0_id, cb_in3_id, num_cores, grid_size](
            const void* operation,
            tt::tt_metal::Program& program,
            const std::vector<Tensor>& input_tensors,
            const std::vector<std::optional<const Tensor>>& optional_input_tensors,
            const std::vector<Tensor>& output_tensors) {
            auto in0_buffer = input_tensors.at(0).buffer();
            auto& mask_tensor = optional_input_tensors.at(0);
            auto out_buffer = output_tensors.size() == 1 ? output_tensors.at(0).buffer() : in0_buffer;

            UpdateDynamicCircularBufferAddress(program, cb_in0_id, *in0_buffer);
            UpdateDynamicCircularBufferAddress(program, cb_out0_id, *out_buffer);
            if (mask_tensor.has_value() && mask_tensor->is_sharded()) {
                UpdateDynamicCircularBufferAddress(program, cb_in3_id.value(), *mask_tensor->buffer());
            }

            if (mask_tensor.has_value()) {
                for (uint32_t i = 0; i < num_cores; ++i) {
                    CoreCoord core = {i % grid_size.x, i / grid_size.x};
                    auto& runtime_args = GetRuntimeArgs(program, reader_kernels_id, core);
                    runtime_args[2] = mask_tensor->buffer()->address();
                }
            }
        };
```

**Comment by yugaoTT** [comment]
oh actually saw that there's set_or_update_runtime_arguments in Batch Norm, I guess you want to cache the tensor memory config and dtype, but not the tensor shape? in that case it's fine.

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/14852904504 - passed
